src/Data/Form.purs

```diff
@@ -127,6 +137,121 @@ instance decodeInputSource :: (DecodeJson a) => DecodeJson (InputSource a) where
+
+  evalField :: Field Expr -> Either EvalError (Field ExprType)
+  evalField field = do
+    description <- evalExpr get field.description
+    input <- evalInput field.input
+    name <- evalExpr get field.name
+    visibility <- evalExpr get field.visibility
+    pure { description, key: field.key, input, name, visibility }
+
+  evalInput :: Input Expr -> Either EvalError (Input ExprType)
+  evalInput = case _ of
+    Dropdown input -> do
+      default <- traverse (evalExpr get) input.default
+      options <- evalExpr get input.options
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- traverse (evalExpr get) input.value
+      pure
+        ( Dropdown
+          { default
+          , options
+          , placeholder
+          , required
+          , value
+          }
+        )
+    Text input -> do
+      default <- traverse (evalExpr get) input.default
+      maxLength <- traverse (evalExpr get) input.maxLength
+      minLength <- traverse (evalExpr get) input.minLength
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- traverse (evalExpr get) input.value
+      pure
+        ( Text
+          { default
+          , maxLength
+          , minLength
+          , placeholder
+          , required
+          , value
+          }
+        )
+    Toggle input -> do
+      default <- traverse (evalExpr get) input.default
+      value <- traverse (evalExpr get) input.value
+      pure (Toggle { default, value })
+
+keys :: Page Expr -> Map Key ExprType
+keys page = foldMap keysSection page.contents
+  where
+  keysSection :: Section Expr -> Map Key ExprType
+  keysSection section = foldMap keysField section.contents
+
+  keysField :: Field Expr -> Map Key ExprType
+  keysField field = case value of
+    Just expr -> case evalExpr (const Nothing) expr of
+      Left _ -> mempty
+      Right x -> Data.Map.singleton field.key x
+    Nothing -> mempty
+    where
+    value = case field.input of
+      Dropdown dropdown -> getValue dropdown
+      Text text -> getValue text
+      Toggle toggle -> getValue toggle
+
+getValue
+  :: ∀ a r
+   . Record (SharedRows a r)
+  -> Maybe a
+getValue x = userInput x.value <|> x.default
+
+setValue :: Key -> ExprType -> Page Expr -> Page Expr
+setValue key val page = page { contents = map setSection page.contents}
+  where
+  setSection :: Section Expr -> Section Expr
+  setSection section = section { contents = map setField section.contents}
+
+  setField :: Field Expr -> Field Expr
+  setField field
+    | key == field.key = case field.input of
+      Dropdown input ->
+        field { input = Dropdown input { value = UserInput (Val val) } }
+      Text input ->
+        field { input = Text input { value = UserInput (Val val) } }
+      Toggle input ->
+        field { input = Toggle input { value = UserInput (Val val) } }
+    | otherwise = field
```

Comment:
Was thinking of how to make this more future-safe. Maybe rewrite as:
```purescript
setField field = if key == field.key
  then case field.input of
     Dropdown input ->
        field { input = Dropdown input { value = UserInput (Val val) } }
      Text input ->
        field { input = Text input { value = UserInput (Val val) } }
      Toggle input ->
        field { input = Toggle input { value = UserInput (Val val) } }
  else field
```
(Not married to the formatting though if you have preferences of when to put things on a new line 😉)

Reply:
I might be missing something, but that seems to be the same as the current code. Sorry for being dense, but which parts change?

Reply:
Ah no, I'm the dense one here. Upon first review I thought the casing was inside the guard, which would mean it didn't have to be total and new `Input` types could fall through to the `otherwise` case. I misread, kindly ignore.

Reply:
No worries.